### PR TITLE
Add SovietsCloset extractor

### PR DESCRIPTION
**Pull request by ChillingPepper**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

[SovietsCloset](https://sovietscloset.com/) is the official stream VOD archive of the streamer SovietWomble, see https://www.twitch.tv/SovietWomble/about.
I added all fields that might make sense, please let me know if there are any mistakes. `uploader`, `creator`, `channel` all seem to be the same for this extractor, is that correct?
I'm also not sure about leading underscores, please advise.

### CODE REVIEW

File: yt_dlp/extractor/sovietscloset.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+
+    def video_meta(self, video_id, stream_date, game_name, catgory_name, episode_number):
+        title = f'SovietWomble {stream_date} - {game_name}'
+        if catgory_name != 'Misc':
+            title += f' - {catgory_name}'
+        title += f' #{episode_number}'
+
+        stream_date_parsed = datetime.datetime.strptime(f'{stream_date} GMT', '%Y-%m-%d %H:%M %Z')
+        stream_date_yyyymmdd = stream_date_parsed.strftime('%Y%m%d')
+        timestamp = int(stream_date_parsed.timestamp())
```
Comment (coletdjnz) [inline yt_dlp/extractor/sovietscloset.py:37]:
```suggestion
        upload_date = unified_strdate(stream_date)
        timestamp = unified_timestamp(stream_date)
```
from utils.py

the timestamp tests fail now but I don't think it was right to begin with (e.g. for the 'SovietWomble 2017-04-13 13:53 - The Witcher #13' test, your timestamp is giving  2017-04-13 11:53) :confused:

File: yt_dlp/extractor/sovietscloset.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+        for key, val in args.items():
+            if val in ('undefined', 'void 0'):
+                args[key] = 'null'
+
+        return self._parse_json(js_to_json(js, args), video_id)['data'][0]
+
+    def video_meta(self, video_id, stream_date, game_name, catgory_name, episode_number):
+        title = f'SovietWomble {stream_date} - {game_name}'
+        if catgory_name != 'Misc':
+            title += f' - {catgory_name}'
```
Comment (coletdjnz) [inline yt_dlp/extractor/sovietscloset.py:32]:
```suggestion
    def video_meta(self, video_id, stream_date, game_name, category_name, episode_number):
        title = f'SovietWomble {stream_date} - {game_name}'
        if category_name != 'Misc':
            title += f' - {category_name}'
```

File: yt_dlp/extractor/sovietscloset.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+
+class SovietsClosetExtractor(InfoExtractor):
+    MEDIADELIVERY_REFERER = {'Referer': 'https://iframe.mediadelivery.net/'}
+
+    def parse_nuxt_jsonp(self, nuxt_jsonp_url, video_id, name):
+        NUXT_JSONP_RE = r'__NUXT_JSONP__\(.*?\(function\((?P<arg_keys>.*?)\)\{return\s(?P<js>\{.*?\})\}\((?P<arg_vals>.*?)\)'
+        nuxt_jsonp = self._download_webpage(nuxt_jsonp_url, video_id, note=f'Downloading {name} __NUXT_JSONP__')
+
+        js, arg_keys, arg_vals = self._search_regex(NUXT_JSONP_RE, nuxt_jsonp, '__NUXT_JSONP__', group=['js', 'arg_keys', 'arg_vals'])
```
Comment (coletdjnz) [inline yt_dlp/extractor/sovietscloset.py:17]:
```suggestion
        nuxt_jsonp = self._download_webpage(nuxt_jsonp_url, video_id, note=f'Downloading {name} __NUXT_JSONP__')

        js, arg_keys, arg_vals = self._search_regex(
            r'__NUXT_JSONP__\(.*?\(function\((?P<arg_keys>.*?)\)\{return\s(?P<js>\{.*?\})\}\((?P<arg_vals>.*?)\)',
            nuxt_jsonp, '__NUXT_JSONP__', group=['js', 'arg_keys', 'arg_vals'])
```

File: yt_dlp/extractor/sovietscloset.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+import datetime
+
+from ..utils import js_to_json
+from .common import InfoExtractor
+
+
+class SovietsClosetExtractor(InfoExtractor):
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/47c7c982aee0296371febde1d862aa324f21b2f2.

File: yt_dlp/extractor/sovietscloset.py (hunk 2 above)
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/5b23de60ce359763b8e6f2258150b7921945add8.

File: yt_dlp/extractor/sovietscloset.py (hunk 3 above)
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/9abcee423035ec58a2b5f3152bab0b3c6fcd4e18.

File: yt_dlp/extractor/sovietscloset.py (hunk 1 above)
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py:38]: This is timezone related. I'm in +02:00 and used my local timezone. I also made the wrong assumption that it's always GMT, but there is summertime, too. Are there any helpers for location based timezones? In this case I'd like to use London.

**Comment by ChillingPepper** [comment]
Are there any guidelines for underscore prefixes?

File: yt_dlp/extractor/sovietscloset.py (hunk 1 above)
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: There doesn't seem to be a way to do that without relying on platform specifics or third-party packages, so I just fixed it per your suggestion in https://github.com/yt-dlp/yt-dlp/pull/884/commits/41f0ea806fec6fa402094f1a1161baa83bca800c.

File: yt_dlp/extractor/sovietscloset.py [hunk 5]
```diff
@@ -0,0 +1,219 @@
...
+        stream = self.parse_nuxt_jsonp(f'{static_assets_base}/video/{video_id}/payload.js', video_id, 'video')['stream']
+
+        return {
+            **self.video_meta(video_id, stream['date'], stream['game']['name'], stream['subcategory']['name'], stream['number']),
+            **self._extract_bunnycdn_iframe(video_id, stream['bunnyId']),
+        }
+
+
+class SovietsClosetPlaylistIE(SovietsClosetBaseIE):
+    _VALID_URL = r'https?://(?:www\.)?sovietscloset\.com/(?!video)(?P<id>.*)'
```
Comment (pukkandan) [inline yt_dlp/extractor/sovietscloset.py:158]:
```suggestion
    _VALID_URL = r'https?://(?:www\.)?sovietscloset\.com/(?!video)(?P<id>[^#?]+)'
```

File: yt_dlp/extractor/sovietscloset.py [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+                playlist_title = game['name']
+                for category in game['subcategories']:
+                    if category['slug'].lower() == category_slug:
+                        if category_slug != 'misc':
+                            playlist_title += ' - ' + category['name']
+                        for stream in category['streams']:
+                            entries.append({
+                                **self.url_result(f'https://sovietscloset.com/video/{stream["id"]}', ie=SovietsClosetIE.ie_key()),
+                                **self.video_meta(stream['id'], stream['date'], game['name'], category['name'], len(entries) + 1),
+                            })
```
Comment (pukkandan) [inline yt_dlp/extractor/sovietscloset.py:218]:
```suggestion
        game = next(game for game in sovietscloset if game['slug'].lower() == game_slug)
        category = next(cat for cat in game['subcategories'] if cat['slug'].lower() == category_slug)
        playlist_title = game.get('name') or game_slug
        if category_slug != 'misc':
            playlist_title += f' - {category.get("name") or category_slug}'
        entries = [{
            **self.url_result(f'https://sovietscloset.com/video/{stream["id"]}', ie=SovietsClosetIE.ie_key()),
            **self.video_meta(stream['id'], stream['date'], game['name'], category['name'], i + 1),
        } for i, stream in enumerate(category['streams'])]
```

**Review by pukkandan — CHANGES_REQUESTED** [review]
Someone more familiar with js than me will need to look into the changes to `js_to_json`

File: yt_dlp/extractor/sovietscloset.py [hunk 7]
```diff
@@ -0,0 +1,219 @@
...
+
+class SovietsClosetBaseIE(InfoExtractor):
+    MEDIADELIVERY_REFERER = {'Referer': 'https://iframe.mediadelivery.net/'}
+
+    def parse_nuxt_jsonp(self, nuxt_jsonp_url, video_id, name):
+        nuxt_jsonp = self._download_webpage(nuxt_jsonp_url, video_id, note=f'Downloading {name} __NUXT_JSONP__')
+        js, arg_keys, arg_vals = self._search_regex(
+            r'__NUXT_JSONP__\(.*?\(function\((?P<arg_keys>.*?)\)\{return\s(?P<js>\{.*?\})\}\((?P<arg_vals>.*?)\)',
+            nuxt_jsonp, '__NUXT_JSONP__', group=['js', 'arg_keys', 'arg_vals']
+        )
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/835ace045d8fbb44e0e6089fe58ddaf0cc027d2a.

File: yt_dlp/extractor/sovietscloset.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+
+        arg_keys = arg_keys.split(',')
+        arg_vals = arg_vals.split(',')
+        args = {key: val for key, val in zip(arg_keys, arg_vals)}
+
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/835ace045d8fbb44e0e6089fe58ddaf0cc027d2a.

File: yt_dlp/extractor/sovietscloset.py [hunk 9]
```diff
@@ -0,0 +1,219 @@
...
+            video_id, note='Downloading BunnyCDN iframe', headers=self.MEDIADELIVERY_REFERER
+        )
+
+        m3u8_url = self._search_regex(r'(https?://.*?\.m3u8)', iframe, 'm3u8 url')
+        thumbnail_url = self._search_regex(r'(https?://.*?thumbnail\.jpg)', iframe, 'thumbnail url')
+
+        m3u8_formats = self._extract_m3u8_formats(m3u8_url, video_id, headers=self.MEDIADELIVERY_REFERER)
+        self._sort_formats(m3u8_formats)
+
+        # TODO duration
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: It was left over, it's removed in https://github.com/yt-dlp/yt-dlp/pull/884/commits/835ace045d8fbb44e0e6089fe58ddaf0cc027d2a.

File: yt_dlp/extractor/sovietscloset.py (hunk 6 above)
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/835ace045d8fbb44e0e6089fe58ddaf0cc027d2a.

File: yt_dlp/extractor/sovietscloset.py [hunk 10]
```diff
@@ -0,0 +1,219 @@
...
+                'episode_number': 3,
+            },
+        },
+    ]
+
+    def _extract_bunnycdn_iframe(self, video_id, bunnycdn_id):
+        iframe = self._download_webpage(
+            f'https://iframe.mediadelivery.net/embed/5105/{bunnycdn_id}',
+            video_id, note='Downloading BunnyCDN iframe', headers=self.MEDIADELIVERY_REFERER
+        )
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/1c1c8a46fcb31f0ecbcb2b28d6e38fe58e6dd250.

File: yt_dlp/extractor/sovietscloset.py [hunk 11]
```diff
@@ -0,0 +1,219 @@
...
+            'id': video_id,
+            'title': title,
+            'http_headers': self.MEDIADELIVERY_REFERER,
+            'uploader': 'SovietWomble',
+            'creator': 'SovietWomble',
+            'release_timestamp': timestamp,
+            'release_date': upload_date,
+            'timestamp': timestamp,
+            'upload_date': upload_date,
+            'uploader_id': 'SovietWomble',
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/d66b454705143131767b5c79dded1dda96c51ca1.

File: yt_dlp/extractor/sovietscloset.py [hunk 12]
```diff
@@ -0,0 +1,219 @@
...
+            'uploader_url': 'https://www.twitch.tv/SovietWomble',
+            'channel': 'SovietWomble',
+            'channel_id': 'SovietWomble',
+            'channel_url': 'https://www.twitch.tv/SovietWomble',
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/d66b454705143131767b5c79dded1dda96c51ca1.

File: yt_dlp/extractor/sovietscloset.py [hunk 13]
```diff
@@ -0,0 +1,219 @@
...
+        args = {key: val for key, val in zip(arg_keys, arg_vals)}
+
+        for key, val in args.items():
+            if val in ('undefined', 'void 0'):
+                args[key] = 'null'
+
+        return self._parse_json(js_to_json(js, args), video_id)['data'][0]
+
+    def video_meta(self, video_id, stream_date, game_name, category_name, episode_number):
+        title = f'SovietWomble {stream_date} - {game_name}'
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/d66b454705143131767b5c79dded1dda96c51ca1.

File: yt_dlp/extractor/sovietscloset.py (hunk 5 above)
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/e8d40b816f12082eca8bb5d4afc08d634ec6e7c2.

File: yt_dlp/extractor/sovietscloset.py [hunk 14]
```diff
@@ -0,0 +1,219 @@
...
+        video_id = self._match_id(url)
+        webpage = self._download_webpage(url, video_id)
+
+        static_assets_base = self._search_regex(r'staticAssetsBase:\"(.*?)\"', webpage, 'staticAssetsBase')
+        static_assets_base = f'https://sovietscloset.com{static_assets_base}'
+
+        stream = self.parse_nuxt_jsonp(f'{static_assets_base}/video/{video_id}/payload.js', video_id, 'video')['stream']
+
+        return {
+            **self.video_meta(video_id, stream['date'], stream['game']['name'], stream['subcategory']['name'], stream['number']),
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/f9669a590b3eecbdc49cba720ca6fba0b4d72c30.

File: yt_dlp/extractor/sovietscloset.py [hunk 15]
```diff
@@ -0,0 +1,219 @@
...
+        m3u8_formats = self._extract_m3u8_formats(m3u8_url, video_id, headers=self.MEDIADELIVERY_REFERER)
+        self._sort_formats(m3u8_formats)
+
+        # TODO duration
+
+        return {
+            'url': m3u8_url,
+            'formats': m3u8_formats,
+            'thumbnail': thumbnail_url,
+        }
```
[the remark this answers is not in the document]
Comment (ChillingPepper) [reply yt_dlp/extractor/sovietscloset.py]: Done in https://github.com/yt-dlp/yt-dlp/pull/884/commits/f9669a590b3eecbdc49cba720ca6fba0b4d72c30.

**Comment by ChillingPepper** [comment]
> Someone more familiar with js than me will need to look into the changes to `js_to_json`

https://github.com/yt-dlp/yt-dlp/pull/884/commits/2d53e1f79f1f0567e2d6d287e795175b61d09ae1 adds support for `$` in identifiers. See https://developer.mozilla.org/en-US/docs/Glossary/Identifier. jQuery famously uses `$` as a shorthand.

https://github.com/yt-dlp/yt-dlp/pull/884/commits/ce42382a3c0ecb6598694d50d7f1afbde8141910 adds support for `undefined` values by translating `undefined` to `null` in JSON. It also recognizes `void 0` as `undefined` (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/void#description), this is used in JS minification.

**Comment by nao20010128nao** [comment]
I think the changes to `js_to_json` is considered valid, though `void` operator accepts any value. (e.g. `void $` in JS is valid and evaluated to `undefined` as well)

**Comment by ChillingPepper** [comment]
This only took a day, very impressive! Thanks all!

**Pull request merged**